components/service/contile/src/main/java/mozilla/components/service/contile/ContileTopSitesProvider.kt

```diff
@@ -0,0 +1,84 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.service.contile
+
+import mozilla.components.concept.fetch.Client
+import mozilla.components.concept.fetch.Request
+import mozilla.components.concept.fetch.isSuccess
+import mozilla.components.feature.top.sites.TopSite
+import mozilla.components.feature.top.sites.TopSitesProvider
+import mozilla.components.support.base.log.logger.Logger
+import mozilla.components.support.ktx.android.org.json.asSequence
+import org.json.JSONException
+import org.json.JSONObject
+import java.io.IOException
+
+internal const val CONTILE_ENDPOINT_URL = "https://contile.services.mozilla.com/v1/tiles"
+
+/**
+ * Provide access to the Contile services API.
+ *
+ * @property client [Client] used for interacting with the Contile HTTP API.
+ */
+class ContileTopSitesProvider(
+    private val client: Client
+) : TopSitesProvider {
+
+    private val logger = Logger("ContileTopSitesProvider")
+
+    @Throws(IOException::class)
+    override suspend fun getTopSites(): List<TopSite.Provided> {
+        return try {
+            fetchTopSites()
+        } catch (e: IOException) {
+            logger.error("Failed to fetch contile top sites", e)
+            throw e
+        }
+    }
+
+    private fun fetchTopSites(): List<TopSite.Provided> {
+        client.fetch(
+            Request(url = CONTILE_ENDPOINT_URL)
+        ).use { response ->
+            if (response.isSuccess) {
+                val responseBody = response.body.string(Charsets.UTF_8)
+
+                return try {
+                    JSONObject(responseBody).getTopSites()
+                } catch (e: JSONException) {
+                    throw IOException(e)
```

Comment:
For `JSONObject.toTopSite()` we return null on exception, but for `JSONObject.getTopSites()` we throw here - it's not clear why we would want to throw it on only this one and then catch it to re-throw it above because the consuming app now needs to try/catch their call as well in order to be safe.

Might be prudent to log the exception and return null/emptyList so that we crash the app if the request is successful but the response has nothing in it.